src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -244,6 +246,233 @@ internal SourceText TryReadFileContent(CommandLineSourceFile file, IList<Diagnos
+
+                allRegexes.Add(config, builder.ToImmutableAndFree());
+            }
+
+            Debug.Assert(allRegexes.Count == analyzerConfigs.Count);
+
+            var optionsBuilder = ImmutableDictionary.CreateBuilder<string, ReportDiagnostic>(
+                CaseInsensitiveComparison.Comparer);
+            foreach (var sourceFile in sourcePaths)
+            {
+                var normalizedPath = PathWithForwardSlashSeparators(sourceFile);
+                // The editorconfig paths are sorted from shortest to longest, so matches
+                // are resolved from most nested to least nested, where last setting wins
+                foreach (EditorConfig config in analyzerConfigs)
+                {
+                    if (normalizedPath.StartsWith(config.Directory, StringComparison.Ordinal))
+                    {
+                        string relativePath = normalizedPath.Substring(config.Directory.Length);
+                        ImmutableArray<Regex> regexes = allRegexes[config];
+                        for (int sectionIndex = 0; sectionIndex < regexes.Length; sectionIndex++)
+                        {
+                            if (regexes[sectionIndex].IsMatch(relativePath))
+                            {
+                                var section = config.NamedSections[sectionIndex];
+                                addDiagnosticOptions(section, optionsBuilder);
+                            }
+                        }
+                    }
+                }
+
+                // Avoid extra allocations by using null.
+                allDiagnosticOptions.Add(optionsBuilder.Count > 0 ? optionsBuilder.ToImmutable() : null);
+                optionsBuilder.Clear();
+            }
+
+            allRegexes.Free();
+            Debug.Assert(allDiagnosticOptions.Count == sourcePaths.Count);
+
+            return allDiagnosticOptions.ToImmutableAndFree();
+
+            void addDiagnosticOptions(
+                EditorConfig.Section section,
+                ImmutableDictionary<string, ReportDiagnostic>.Builder builder)
+            {
+                const string DiagnosticOptionPrefix = "dotnet_diagnostic.";
+                const string DiagnosticOptionSuffix = ".severity";
+
+                foreach (var (key, value) in section.Properties)
+                {
+                    // Keys are lowercased in editorconfig parsing
+                    if (key.StartsWith(DiagnosticOptionPrefix, StringComparison.Ordinal) &&
+                        key.EndsWith(DiagnosticOptionSuffix, StringComparison.Ordinal))
+                    {
+                        var diagId = key.Substring(
+                            DiagnosticOptionPrefix.Length,
+                            key.Length - (DiagnosticOptionPrefix.Length + DiagnosticOptionSuffix.Length));
+
+                        ReportDiagnostic? severity;
+                        if (CaseInsensitiveComparison.Equals(value, "default"))
+                        {
```

Comment:
This can be done using a switch on value.ToLowerInvariant(). Since we're looking at ASCII, no fancy Unicode-correct case insensitivity is needed.

Why do we require an exact match for the prefix and suffix of the key, but compare the middle part using relaxed case?

Reply:
I've changed this to use `OrdinalIgnoreCase`. Jared suggested not calling ToLower or ToUpper because it will allocate an additional string for each property. The `if` is more verbose, but avoids that.

As to the mixed casing -- the key is always lower-cased by the EditorConfig parser, while the value is case-preserving, so we compare the value case-insensitively, but the key case-sensitively.